pallets/parachain-staking/src/lib.rs

```diff
@@ -1650,6 +1652,60 @@ pub mod pallet {
 			<SelectedCandidates<T>>::put(collators);
 			(collator_count, delegation_count, total)
 		}
+
+		/// Apply the delegator intent for revoke and decrease in order to build the
+		/// effective list of delegators with their intended bond amount.
+		///
+		/// This is will:
+		/// - if [DelegationChange::Revoke] is outstanding, set the bond amount to 0.
+		/// - if [DelegationChange::Decrease] is outstanding, subtract the bond by specified amount.
+		/// - else, do nothing
+		///
+		/// The intended bond amounts will be used while calculating rewards.
+		fn get_rewardable_delegators(
+			collator: &T::AccountId,
+		) -> Vec<Bond<T::AccountId, BalanceOf<T>>> {
+			<TopDelegations<T>>::get(collator)
+				.expect("all members of CandidateQ must be candidates")
+				.delegations
+				.into_iter()
+				.map(|mut bond| {
+					let delegator = <DelegatorState<T>>::get(&bond.owner)
+						.expect("delegator state must be present for a bonded delegation");
```

Comment:
This adds a read for every delegation that is included in the CollatorSnapshots at each round transition. It is an additional `64 (total_selected) * 300 (top_delegation_len)` reads at each round transition so benchmarking needs to be updated accordingly, but that can be done in a followup ticket.

Reply:
This could be a lot in terms of PoV because `DelegatorState` contains all delegations and requests. It needs to be split into metadata, delegations, requests as described in MOON 1623

Reply:
Oh, actually, @4meta5 is right, this is going to introduce a lot of reads and have a big impact on PoV during round changes. We're essentially looking up the `DelegatorState` for each active delegation during a round change.

We should be careful with this change because it will easily be a bottleneck.

Reply:
@4meta5 actually the net increase is  `300 (top_delegation_len)` since the `64 (total_selected)` was already there. The new functionality adds the additional delegation read per top delegation. I'd state the complexity as subquadratic `O(c*d)` in terms of total selected collator count, and linear `O(d)` in terms of top delegator count.

Splitting the Delegator State is certainly helpful and should be done, however even in that case, we'd probably have multiple writes per operation in distinct storages. We'd need to maintain a delegation request per collator if we are to leverage a single read per collator reward distribution, which might not work with the current design of DelegatorState being indexed by delegator id. Using a double map would lead to the same number of reads again.

IMO the only "free" read we will ever get during reward distribution is from the `Bond` struct (provided we mutate it as the requests come along), but I take we don't wish to add additional fields to the `Bond` struct.
The only other alternate is to unwrap the Delegator State into and store them

Reply:
That looks to be way too heavy. If I understand correctly, we are moving from a O(1) read from the top delegations to O(n). This would be too big for the PoV.
The only case where doing O(n) would be acceptable is if the `n` was only the delegations with a request (which should be small). This could be achieved by maintaining in the top delegation if a delegator has a request pending or not for exemple

Reply:
I agree, seems we need to mirror the delegator requests for now on `DelegatorState` and a new `TopRewardableDelegators` which seems could be a map from collator id to `BTreeMap<delegator_id, bond_amount>`

Reply:
I think the problem lies in the way we store the `PendingDelegationRequests`.
I think we need to remove the `requests` field from the `Delegator` struct and create a double map `PendingDelegationRequestsPerDelegatorPerCollator`, with the delegator as the 1st key and the collator as the 2nd key.
This would make the PoV much lighter!

Reply:
I'd be interested in knowing, how does using a double map can improve the PoV size?
As far as I could tell, double map doesn't really help much in reducing the amount of reads (couldn't find a way to get all objects under `k1` without an `iter()`) - so that's still a read per delegator to check for its requests.

Reply:
It's still a read per delegator but you're only reading 1 request instead of reading all the requests. That's why it's a smaller PoV.

I like @librelois 's suggestion and think we should go with that. It is moving requests out of DelegatorState so will require a migration.

Reply:
The PoV contains the entirety of the value associated with the key read, even if only a part of this value is used.
A double map allows to retrieve only the unique pending request associated to a couple (delegator, collator), and only if this query exists.
While the current code writes in the PoV the whole `DelegatorState` of all delegators.
Indeed the number of reads would remain the same, but the PoV would be smaller.

If we want to reduce the number of reads, we would have to store the pending delegator requests in 2 places (to get them by collator here, and by delegator at runtime), but this would imply more writes, and I think it's better to minimize the writes.

Reply:
Yeah that would work with the original concept of the double map then, the reads will remain same, but the size of the PoV would be reduced.

Reply:
Proposal:
```rust
	#[derive(Clone, Eq, PartialEq, Encode, Decode, RuntimeDebug, TypeInfo)]
        pub enum DelegationChangeAction<Balance> {
		Revoke,
		Decrease(Balance),
	}

	#[derive(Clone, Eq, PartialEq, Encode, Decode, RuntimeDebug, TypeInfo)]
	pub struct PendingRequest<Balance> {
		pub when_executable: RoundIndex,
		pub action: DelegationChangeAction<Balance>,
	}

	#[pallet::storage]
	#[pallet::getter(fn delegator_pending_requests)]
	pub(crate) type DelegatorPendingRequests<T: Config> = StorageDoubleMap<
		_,
		Twox64Concat,
		T::AccountId,  // delegator-id
		Twox64Concat,
		T::AccountId, // collator-id
		PendingRequest<BalanceOf<T>>,
		OptionQuery,
	>;
```

This can remove some of the unneeded fields like `collator` & `amount` and move them at the point of usage.